### PR TITLE
[YUNIKORN-36] Cleanup leaked resources when deleting the helm chart

### PR DESCRIPTION
**Pull request by TaoYang526**
Please refer to issue: https://issues.apache.org/jira/browse/YUNIKORN-36

### CODE REVIEW

**Comment by wilfred-s** [comment]
I don't think that relying on hyperkube is the right thing as kubernetes has just merged the changes to stop releasing the hyperkube container image [kubernetes/kubernetes#88676](https://github.com/kubernetes/kubernetes/pull/88676)

**Comment by yangwwei** [comment]
Hi @TaoYang526 

Can we leverage the yunikorn scheduler docker image which already has the script installed, we can do 
```
admission_util.sh delete
```
to clean up everything

**Comment by TaoYang526** [comment]
Thanks @wilfred-s , @yangwwei  for the review.
> Can we leverage the yunikorn scheduler docker image which already has the script installed, we can do `admission_util.sh delete ` to clean up everything

I think we can't do that since the command will be called in PreStop hook which is called immediately before a container is terminated due to an API request or management event such as liveness probe failure, preemption, resource contention and others, not just before the final termination. (Please refer to https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/#container-hooks)

I haven't find a proper image to replace deprecated hyperkube image after a quick search, I'll find another image or another way to solve this problem. Please share your thoughts if you have any idea about this. Thanks.

**Comment by TaoYang526** [comment]
Hi, @yangwwei @wilfred-s 
This PR has been updated to clean up things via curling to APIServer instead of relying on hyperkube which is removed from k8s base images, please help to review it again. Thanks.

**Comment by TaoYang526** [comment]
Thanks @yangwwei for the feedback about the URL of APIServer.
I discovered that there's something wrong in the former image which may cause DNS can't resolve the service address correctly in search domains defined in /etc/resolv.conf, that's why it had to use the full service name with suffix '.cluster.local' to connect with API server.
Latest commit replaces the incorrect curl image with official curl image, and update the URL of APIServer with non-suffix service name: 'kubernetes.default.svc' just as described in the kubernetes official doc: https://kubernetes.io/docs/tasks/administer-cluster/access-cluster-api/#accessing-the-kubernetes-api.

File: helm-charts/yunikorn/templates/cleanup.yaml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+            - >
+              APISERVER=https://kubernetes.default.svc;
+              SERVICEACCOUNT=/var/run/secrets/kubernetes.io/serviceaccount;
+              NAMESPACE=$(cat ${SERVICEACCOUNT}/namespace);
+              TOKEN=$(cat ${SERVICEACCOUNT}/token);
+              CACERT=${SERVICEACCOUNT}/ca.crt;
+              curl --cacert ${CACERT} --header "Authorization: Bearer ${TOKEN}" -X DELETE ${APISERVER}/api/v1/namespaces/${NAMESPACE}/configmaps/yunikorn-configs;
+              curl --cacert ${CACERT} --header "Authorization: Bearer ${TOKEN}" -X DELETE ${APISERVER}/api/v1/namespaces/${NAMESPACE}/serviceaccounts/yunikorn-admin;
+              curl --cacert ${CACERT} --header "Authorization: Bearer ${TOKEN}" -X DELETE ${APISERVER}/apis/rbac.authorization.k8s.io/v1beta1/clusterrolebindings/yunikorn-rbac;
+      restartPolicy: OnFailure
```
Comment (yangwwei) [inline helm-charts/yunikorn/templates/cleanup.yaml:38]: I feel we don't need to restart this on failure, if the execution failed, probably the second attempt will fail too.
Can we change this to Never?

File: helm-charts/yunikorn/templates/cleanup.yaml [hunk 2]
```diff
@@ -0,0 +1,38 @@
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: yunikorn-cleanup
+  namespace: {{ .Release.Namespace }}
+  annotations:
+    "helm.sh/hook": post-delete
+    "helm.sh/hook-weight": "3"
```
Comment (yangwwei) [inline helm-charts/yunikorn/templates/cleanup.yaml:8]: My understanding the weight is giving us a determinable order when installing resources.
But why we have the weight=3 for the cleanup?
This is a post-delete hook, which only triggers when we delete the helm chart, correct?

Comment (TaoYang526) [reply helm-charts/yunikorn/templates/cleanup.yaml:8]: Yes, absolutely right. I will remove this annotation. Thanks for this reminder.

File: helm-charts/yunikorn/templates/cleanup.yaml (hunk 1 above)
Comment (TaoYang526) [reply helm-charts/yunikorn/templates/cleanup.yaml:38]: Make sense.

**Comment by TaoYang526** [comment]
Thanks @yangwwei for the review.

> With this patch, does it mean we can delete all yunikorn created resources by a simple "helm delete" command?  do we still need the pre-stop hook?

We can clean up all resources that yunikorn created by "helm uninstall" command. However, I think the pre-stop hook is more suitable to do that for admission-controller, since the resources created for admission-controller should better be managed by admission-controller itself, otherwise if these resources changed we may need to update cleanup.yaml for helm-charts as well as admission_util.sh script for admission-controller. Thoughts?

**Comment by yangwwei** [comment]
I did some tests, install/delete helm chart again and again.

I found an issue we need to fix. When I delete the helm-chart second time, the job fails because the job (with that name) already exists. I think we can fix this by

```
-  name: yunikorn-cleanup
+  generateName: yunikorn-cleanup
```

so every time the job will have a different name.

Another thing is when the job finished, the job is not deleted and the cleanup pod is left-over. Can we take a look at if https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/#ttl-mechanism-for-finished-jobs can help the cleanup?

Other than these issues, it looks good.

**Comment by TaoYang526** [comment]
Thanks @yangwwei for the feedback.

> When I delete the helm-chart second time, the job fails because the job (with that name) already exists.

Do you mean that "helm uninstall" command was executed again before the uninstall process was done? I think it's ok since the first execution couldn't be affected by the second one, right?

> when the job finished, the job is not deleted and the cleanup pod is left-over.

I haven't seen this happen, it should be cleaned up after the job is done since we have configured `"helm.sh/hook-delete-policy": hook-succeeded` in cleanup.yaml. Could you please elaborate how this may happen?

**Comment by TaoYang526** [comment]
After discussed with @yangwwei offline, it seems that `"helm.sh/hook-delete-policy": hook-succeeded` config may not works well in helm v2 (at least for several specific versions: https://github.com/helm/helm/issues/1769), we can assume that it can work well in helm v3 and I have already tested it for many times in my local environment. 

Also, I have tested to add generateName but fails to uninstall since it's not available in helm(https://github.com/helm/helm/issues/3348), and TTL may not work well unless that both kube-apiserver and kube-controller-manager enable feature gate TTLAfterFinished according to https://kubernetes.io/docs/concepts/workloads/controllers/ttlafterfinished/.

Now helm v3 is the stable and default version according to helm official website(https://helm.sh/docs/), and there are lots of changes between v2 and v3, we decided to do nothing more for this PR and update helm commands based on helm v3 in the User Guide document via another PR in core project: https://github.com/apache/incubator-yunikorn-core/pull/129

**Comment by yangwwei** [comment]
I agree this sounds good to me. 
We can suggest people use helm3, for which I think most people are already doing that. Let's track the doc changes in another ticket. This one looks good to go. Thanks!

**Comment by yangwwei** [comment]
BTW, this approach is following the doc: https://kubernetes.io/docs/tasks/administer-cluster/access-cluster-api/#without-using-a-proxy

**Pull request merged**